civiform/civiform
# A11y: improve focus management for validation message reporting

### Description

This PR improves focus management in the application UI when there are validation errors. The previous code attempted to do this as well, but an audit in September (and manual testing) revealed it was not working as intended. See https://github.com/civiform/civiform/issues/5519: 

> After displaying the error message, programmatically move the focus to the first input field with an error

Previously, civiform added the `autofocus` HTML attribute to all elements in a question, relying on the browser to focus the first element with the attribute as [described here](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/autofocus).

This approach instead introduces a new enum, `AutoFocusTarget`, that instructs question renderers how to use `autofocus`, either not adding it, adding it to the first input that has an associated error (when reporting errors), or adding it to the first input (when the applicant has navigated directly to a question from the review page).

### Checklist

#### General

Read the full guidelines for PRs [here](https://github.com/civiform/civiform/wiki/Technical-contribution-guide#creating-a-pull-request)


#### User visible changes


### Instructions for manual testing

As an applicant starting an application:

1. Begin a new application
2. Click "continue"
3. No input element should have focus

As an applicant navigating to a specific question in an application:

1. Begin a new application
2. Click on a specific question
3. The first input of the question you clicked should have focus

As an applicant viewing a validation error message:

1. Navigate to a screen with a required question or some other validation
2. Submit the form with incorrect/missing data
3. The first invalid field of the first invalid question should have focus

### Issue(s) this completes

Addresses an item in https://github.com/civiform/civiform/issues/5519
